### ISSUE TITLE
Notify packages when their new versions are released

### ISSUE THREAD

**Issue by christopher-dG** [JuliaRegistries/RegistryCI.jl#93]
This is for https://github.com/JuliaRegistries/TagBot/pull/32.
This is mostly a doc for myself. I'll get to it eventually after I have TagBot itself working the way I like it, but if anyone wants to do it in the meantime, feel free :smile:.

### What needs to happen

When a new version is merged (new packages are included in this), an issue comment should be made in that package's repo, with the contents:

```toml
version = "1.2.3"
sha = "abcdef"
```
- This is TOML syntax
- SHA is the commit hash (not tree) of the version

To be minimally obtrusive, I was thinking to create an issue, immediately close it, and comment there. The issue should have a constant title, so that we can look it up and reuse it for future releases instead of creating new issues every time.
This should happen via GitHub Actions (similar to the automerge), so that `github-actions[bot]` is the user making the comments, which TagBot expects.
Optionally, the issue and comment could be deleted, but I think the user will still get email notifications and it might be even more annoying and confusing if you click an email link and get a 404.

### Why this way?

Here are some very nice alternatives that don't work:

#### Repository dispatch events

[Repository dispatch events](https://help.github.com/en/github/automating-your-workflow-with-github-actions/events-that-trigger-workflows#external-events-repository_dispatch) are basically webhooks that don't require setup, which you can use to trigger a GitHub Action in another repository. However, to invoke a dispatch event in a repo, you must have a token with content write permissions in that repo, such as one obtained via GitHub Apps. Since we're trying to avoid handing over write permissions to random Apps, this won't work.

#### Commit comments

Why not just comment on the commit in the repo instead of creating an issue?
Well, GitHub Actions currently can't respond to commit comments for some reason. See [here](https://github.community/t5/GitHub-Actions/Can-we-have-on-commit-comment-as-a-trigger/td-p/36667) for a feature request.
If it's ever supported, then switching to this method is a no-brainer, except it would require everyone using the issue-comment method to change their workflow files, which would really suck.

**Comment by DilumAluthge** [comment]
Personally, I don’t think this is a good approach.

Note that `github-actions[bot]` is the username used by any GitHub Action on any repository.

So I could set up a repository with a GitHub Action that posts these issue comments to all kinds of Julia packages. And since there is no authentication or verification, I could send all kinds of gibberish and try to trick TagBot into making unnecessary tags. With this system, I could create unlimited tags on any Julia package.

If you want to go the GitHub Actions approach, I think this is how you do it:
1. In the Julia package repository on which TagBot is installed as a GitHub Action, you have a `cron` job. (I believe that GitHub Actions calls them `on:schedule` jobs.)
2. The `cron` job runs, for example, every 10 minutes.
3. When the cron job, we download the latest version of `Versions.toml` from the General registry. For example, if TagBot is installed on the DataFrames.jl repository, then we would download the https://raw.githubusercontent.com/JuliaRegistries/General/master/D/DataFrames/Versions.toml file. Because we are downloading over HTTPS, we have assurance that the content we are getting is truly the content of the General registry.
4. We parse the `Versions.toml` file and make a list of all of the version numbers and tree hashes.
5. We make a list of all of the Git tags that currently exist on our repository.
6. If every version in `Versions.toml` has a corresponding Git tag, we stop.
7. If there are one or more versions in `Versions.toml` that do not have corresponding Git tags, then we need to make those tags. To make a tag: we know the version number and the tree sha. We search commits in reverse chronological order until we find the most recent commit that matches the given tree sha. Then we make tag at that commit for the appropriate version number.
8. We repeat until we have made tags for all of the versions in `Versions.toml` that did not previously have tags.

**Comment by fredrikekre** [comment]
I was also going to suggest a cron-job.

> an issue comment should be made in that package's repo

I don't think we are allowed (i.e. not authorized) to post to other repositories.

**Comment by christopher-dG** [comment]
Cron makes sense, I'm not not that I like the idea of possibly making a bunch of releases at KBE the first time though. Thanks for the feedback!

**Comment by DilumAluthge** [comment]
> I'm not not that I like the idea of possibly making a bunch of releases at KBE the first time though.

That's easy enough to fix.

Clone the General registry over HTTPS. Checkout the master branch, look at the file `D/DataFrame/Versions.toml`, and make a list of all version numbers that exist in `Versions.toml` on the `master` branch.

Then, checkout a commit from a few days ago. For example, go back through the commits and find any commit that is at least `3` days old and no more than `7` days old. (Substitute whatever numbers you want in place of the `3` and `7`.) Checkout that commit. It really doesn't matter which specific commit you use, as long as that commit is at least a few days old. Now, after checking out this old commit, look at the file `D/DataFrame/Versions.toml`, and make a list of all version numbers that exist in `Versions.toml` on this old commit.

Now you have two sets of version numbers:
- The set of all version numbers as of the latest `master` - call this set `A`
- The set of all version numbers as of an old commit a few days old - call this set `B`.

Take the set difference of those two sets. Call this set `C`. In other words, set `C` contains all the elements of set `A` that are not elements of set `B`.

Now, `C` is the set of only the version numbers that were added in the last few days.

If  `C` is empty, then stop.

If  `C` is not empty, then compare that set to the set of existing Git tags.

If, for every version number in `C`, there exists the corresponding Git tag, then stop.

Else, for every version number in `C` for which there does not exist a corresponding Git tag, then create the Git tag using the process I described above.

By using this approach, you will only create Git tags for PR's that were merged into the General registry in the last few days.
